src/os/filestore/FileStore.cc

```diff
@@ -1794,11 +1806,249 @@ int FileStore::umount()
+    error = thr->Start();
+    if (error)
+      break;
+  }
+
+  if (error) {
+    derr << "unable to create read completion threads, error = "
+	 << error << dendl;
+    for (auto rct : read_completion_threads)
+      delete rct;
+    read_completion_threads.clear();
+  }
+ 
+  return error;
+}
+
+void FileStore::destroy_read_completion_threads()
+{
+  if (read_completion_threads.size() == 0)
+    return;
+  dout(0) << "FileStore: destroying " << read_completion_threads.size()
+	  << " read completion threads." << dendl;
+
+  for (auto rct : read_completion_threads)
+    delete rct;
+
+  dout(0) << "FileStore:  destroy_read_completion_threads done" << dendl;
+  read_completion_threads.clear();
+}
+
+int FileStore::async_read_dispatch(
+  Sequencer *posr,
+  Context *ctx,
+  const coll_t& cid,
+  const ghobject_t& oid,
+  uint64_t off,
+  size_t len,
+  bufferlist *bl,
+  uint32_t op_flags,
+  Context *blessed_context)
+{
+  OpSequencer *osr = may_create_op_sequencer(posr);
+  // PGID-to-completion thread mapping:
+  int thread_index = osr->id % read_completion_threads.size();
+
+  FDRef fd;
+  off64_t offset = (off64_t) off;
+  int r = lfn_open(cid, oid, false, &fd, 0, true);
+  if (r < 0) {
+    dout(10) << "FileStore::async_read_dispatch(" << cid << "/" << oid << ") open error: "
+	     << cpp_strerror(r) << dendl;
+    return r;
+  }
+
+  if (len == 0) {
+    struct stat st;
+    memset(&st, 0, sizeof(struct stat));
+    int r = ::fstat(**fd, &st);
+    assert(r == 0);
+    len = st.st_size;
+  }
+
+#ifdef HAVE_POSIX_FADVISE
+  if (op_flags & CEPH_OSD_OP_FLAG_FADVISE_RANDOM)
+    posix_fadvise(**fd, offset, len, POSIX_FADV_RANDOM);
+  if (op_flags & CEPH_OSD_OP_FLAG_FADVISE_SEQUENTIAL)
+    posix_fadvise(**fd, offset, len, POSIX_FADV_SEQUENTIAL);
+#endif
+
+  // Extend to 512 bytes since direct IO is being used
+  unsigned alignment = len % 512;
+  if (alignment)
+    alignment = 512 - alignment;
+
+  read_aio_bl(fd, bl, offset, len, alignment, ctx, blessed_context, op_flags, oid, thread_index);
+  return r;
+}
+
+void FileStore::read_aio_bl(FDRef fdref, bufferlist *bl, off64_t& pos,
+                            size_t len, unsigned alignment, Context *ctx, Context *pctx,
+                            uint32_t op_flags, const ghobject_t& oid, int thread_index)
+{
+  int fd = **fdref;
+  read_aio *rio = new read_aio(fdref, bl, pos, len, alignment, op_flags, oid, ctx, pctx);
 
+  len += alignment;
+  io_prep_pread(&rio->iocb, fd, rio->bptr->c_str(), len, pos);
+  dout(20) << "read_aio_bl .. " << len << dendl;
+
+  ReadCompletionThread& rct = *read_completion_threads[thread_index];
+  // Error if thread no longer up.
+  int status = rct.inc_thrd_pending();
+  if (status < 0) {
+    rio->len = status;
+    async_read_finish(rio);
+    return;
+  }
+
+  io_context_t *iocontext = &rct.read_aio_ctx;
+
+  iocb *piocb = &rio->iocb;
+  int attempts = 10;
+  do {
+    int r = io_submit(*iocontext, (long) 1, &piocb);
```

Comment:
As a matter of fact io_submit can block even if the file has been
opened with O_DIRECT, see 
http://linuxplumbersconf.org/2009/slides/Anthony-Liguori-qemu-block.pdf,
http://blog.libtorrent.org/2012/10/asynchronous-disk-io
This fact is somewhat mentioned in the kernel source.
sys_io_submit -> do_io_submit -> io_submit_one -> aio_run_iocb,
which (for reads) calls a filesystem dependent read_iter.
Documentation/fs/vfs.txt describes read_iter as "possibly
asynchronous read with iov_iter as destination".
http://lxr.free-electrons.com/source/Documentation/filesystems/vfs.txt?v=4.1#L839

For ext4 read_iter is generic_file_read_iter which calls
filemap_write_and_wait_range (which can block).
generic_file_read_iter uses (filesystem specific) direct_IO operation,
which is ext4_direct_IO. ext4_direct_IO delegates reads to ext4_ind_direct_IO,
which can block for many reasons, in particular, in ext4_get_block which can 
block (since it reads the filesystem metadata which might need to wait for 
a journal commit, etc). That said I'm not a Linux hacker, so this explanation
might be well wrong. However it doesn't change the fact that Linux "maybe-AIO"
is really asynchronous only with block devices opened with O_DIRECT.
This is the main reason why IO heavy applications such as qemu, torrent clients, and ceph itself avoid using Linux AIO with ordinary files.